### PR TITLE
[USM] Support all http2 methods

### PR DESCRIPTION
**Pull request by amitslavin**
### What does this PR do?

The PR introduces the support for methods that are not fully exist in the static table.

We're supporting the case of a method being passed as a literal header
instead of a fully indexed entry in the static table.

The PR creates a new struct status_code_t which covers the 2 options we have:

indexed_value - represents a fully indexed in the static table
literal - so we introduce is_huffman_encoded flag, length (which represents the string length of the method) and a buffer to save the raw value.
In the user mode we're converting the correct methods based on our options.

The PR decrease the `HTTP2_BATCH_SIZE` in order to support the method struct with in `http2_stream_t`.

### Motivation

While for the original purpose of gRPC monitoring the methods in the static table suffice, for HTTP2 monitoring (like an implicit upgrade of HTTPs connection) it is a gap, as there is not guarantee the methods are restricted to the list in the static table.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided. Except if the `qa/skip-qa` label, with required either `qa/done` or `qa/no-code-change` labels, are applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/network/ebpf/c/protocols/http2/decoding.h [hunk 1]
```diff
@@ -341,6 +342,11 @@ static __always_inline void process_headers(struct __sk_buff *skb, dynamic_table
                 bpf_memcpy(current_stream->status_code.raw_buffer, dynamic_value->buffer, HTTP2_STATUS_CODE_MAX_LEN);
                 current_stream->status_code.is_huffman_encoded = dynamic_value->is_huffman_encoded;
                 current_stream->status_code.finalized = true;
+            }  else if (is_method_index(dynamic_value->original_index)) {
+                bpf_memcpy(current_stream->request_method.raw_buffer, dynamic_value->buffer, HTTP2_METHOD_MAX_LEN);
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/http2/decoding.h:346]: request_started?

File: pkg/network/protocols/http2/model_linux.go [hunk 2]
```diff
@@ -134,16 +134,56 @@ func (tx *EbpfTx) ConnTuple() types.ConnectionKey {
...
+	case "HEAD":
+		return http.MethodHead, nil
+	case "OPTIONS":
+		return http.MethodOptions, nil
+	case "PATCH":
+		return http.MethodPatch, nil
+	// Add more cases for other HTTP methods as needed
+	default:
+		return 0, fmt.Errorf("unsupported HTTP method: %s", method)
+	}
```
Comment (guyarb) [inline pkg/network/protocols/http2/model_linux.go:153]: take a look in the definition of the Method* consts
eventually something like:
```go
switch vaL := strings.ToUpper(method):
case http.MethodPut, http.MethodDelete,...:
    return val, nil
default:
....
```

Comment (guyarb) [reply pkg/network/protocols/http2/model_linux.go:153]: also, let's add support for post and get, as there is no denial for such use-case to happen

File: pkg/network/protocols/http2/model_linux.go [hunk 3]
```diff
@@ -134,16 +134,56 @@ func (tx *EbpfTx) ConnTuple() types.ConnectionKey {
...
+			return 0
+		}
+		http2Method, err := stringToHTTPMethod(method)
+		if err != nil {
+			return 0
+		}
+		return http2Method
+	}
+
+	method := string(tx.Stream.Request_method.Raw_buffer[:tx.Stream.Request_method.Length])
```
Comment (guyarb) [inline pkg/network/protocols/http2/model_linux.go:181]:
```suggestion
	var method string
	var err error
	if tx.Stream.Request_method.Is_huffman_encoded {
		method, err = hpack.HuffmanDecodeToString(tx.Stream.Request_method.Raw_buffer[:tx.Stream.Request_method.Length])
		if err != nil {
			return 0
		}
	} else {
		method = string(tx.Stream.Request_method.Raw_buffer[:tx.Stream.Request_method.Length])
	}
	http2Method, err := stringToHTTPMethod(method)
	if err != nil {
		return 0
	}
	return http2Method
```

File: pkg/network/usm/usm_http2_monitor_test.go [hunk 4]
```diff
@@ -754,9 +754,9 @@ func (s *usmHTTP2Suite) TestRawTraffic() {
 		},
 		{
 			name: "validate http methods",
-			// The purpose of this test is to validate that we are not supporting http2 methods different from POST and GET.
+			// The purpose of this test is to validate that we are able to capture all http methods.
```
Comment (guyarb) [inline pkg/network/usm/usm_http2_monitor_test.go:757]: let's add tests for Get and Post which are literals

File: pkg/network/usm/usm_http2_monitor_test.go [hunk 5]
```diff
@@ -754,9 +754,9 @@ func (s *usmHTTP2Suite) TestRawTraffic() {
...
 			messageBuilder: func() [][]byte {
-				httpMethods = []string{http.MethodHead, http.MethodPut, http.MethodPatch, http.MethodDelete, http.MethodOptions}
+				httpMethods = []string{http.MethodHead, http.MethodDelete, http.MethodPut, http.MethodPatch, http.MethodOptions}
```
Comment (guyarb) [inline pkg/network/usm/usm_http2_monitor_test.go:759]: let's add all methods (Trace, Connect)

**Review by guyarb — CHANGES_REQUESTED** [review]

File: pkg/network/ebpf/c/protocols/http2/decoding-tls.h [hunk 6]
```diff
@@ -344,6 +345,11 @@ static __always_inline void tls_process_headers(tls_dispatcher_arguments_t *info
                 bpf_memcpy(current_stream->status_code.raw_buffer, dynamic_value->buffer, HTTP2_STATUS_CODE_MAX_LEN);
                 current_stream->status_code.is_huffman_encoded = dynamic_value->is_huffman_encoded;
                 current_stream->status_code.finalized = true;
+            } else if (is_method_index(dynamic_value->original_index)) {
```
[the remark this answers is not in the document]
Comment (amitslavin) [reply pkg/network/ebpf/c/protocols/http2/decoding-tls.h:348]: you are right, nice catch

File: pkg/network/ebpf/c/protocols/http2/decoding.h (hunk 1 above)
Comment (amitslavin) [reply pkg/network/ebpf/c/protocols/http2/decoding.h:346]: you are right, nice catch

File: pkg/network/usm/usm_http2_monitor_test.go (hunk 4 above)
Comment (amitslavin) [reply pkg/network/usm/usm_http2_monitor_test.go:757]: added

File: pkg/network/protocols/http2/model_linux.go [hunk 7]
```diff
@@ -147,38 +147,39 @@ func stringToHTTPMethod(method string) (http.Method, error) {
...
 		default:
-			return http.MethodUnknown
-		}
-	}
-
-	if tx.Stream.Request_method.Is_huffman_encoded {
-		method, err := hpack.HuffmanDecodeToString(tx.Stream.Request_method.Raw_buffer[:tx.Stream.Request_method.Length])
-		if err != nil {
-			return 0
+			method = string(tx.Stream.Request_method.Raw_buffer[:tx.Stream.Request_method.Length])
```
Comment (guyarb) [inline pkg/network/protocols/http2/model_linux.go:180]: the new format does not make sense

Comment (guyarb) [reply pkg/network/protocols/http2/model_linux.go:180]: you're handling the "literal but not encoded" scenario inside the "static table" use-cases

File: pkg/network/usm/usm_http2_monitor_test.go [hunk 8]
```diff
@@ -756,7 +756,10 @@ func (s *usmHTTP2Suite) TestRawTraffic() {
 			name: "validate http methods",
 			// The purpose of this test is to validate that we are able to capture all http methods.
 			messageBuilder: func() [][]byte {
-				httpMethods = []string{http.MethodHead, http.MethodDelete, http.MethodPut, http.MethodPatch, http.MethodOptions}
+				httpMethods = []string{http.MethodGet, http.MethodPost, http.MethodHead, http.MethodDelete,
+					http.MethodPut, http.MethodPatch, http.MethodOptions, "TRACE", "CONNECT"}
```
Comment (guyarb) [inline pkg/network/usm/usm_http2_monitor_test.go:760]: http.MethodTrace, http.MethodConnect

File: pkg/network/usm/usm_http2_monitor_test.go [hunk 9]
```diff
@@ -767,6 +770,14 @@ func (s *usmHTTP2Suite) TestRawTraffic() {
...
 			},
 			expectedEndpoints: map[usmhttp.Key]int{
+				{
+					Path:   usmhttp.Path{Content: usmhttp.Interner.GetString(http2DefaultTestPath)},
+					Method: usmhttp.MethodGet,
+				}: 1,
+				{
+					Path:   usmhttp.Path{Content: usmhttp.Interner.GetString(http2DefaultTestPath)},
+					Method: usmhttp.MethodPost,
+				}: 1,
```
Comment (guyarb) [inline pkg/network/usm/usm_http2_monitor_test.go:780]: that's great, but where do you test the scenario those methods are being sent as literal?
also, you should add tests for encoded versions

File: pkg/network/protocols/http2/model_linux.go (hunk 7 above)
Comment (amitslavin) [reply pkg/network/protocols/http2/model_linux.go:180]: you are right, changed

File: pkg/network/usm/usm_http2_monitor_test.go (hunk 9 above)
Comment (amitslavin) [reply pkg/network/usm/usm_http2_monitor_test.go:780]: added tests for both scenarios

File: pkg/network/usm/usm_http2_monitor_test.go [hunk 10]
```diff
@@ -800,6 +800,100 @@ func (s *usmHTTP2Suite) TestRawTraffic() {
...
+				headerFields := removeHeaderFieldByKey(testHeaders(), ":method")
+				headersFrame, err := usmhttp2.NewHeadersFrameMessage(usmhttp2.HeadersFrameOptions{
+					Headers: headerFields,
+				})
+				require.NoError(t, err, "could not create headers frame")
+				for i, method := range httpMethods {
+					huffMethod := hpack.AppendHuffmanString([]byte{}, method)
+					// we are adding 128 to the length of the huffman encoded method,
+					// as it is the representation of the huffman encoding (MSB ofo the octet is on).
+					rawMethod := append([]byte{0x43}, byte(128+len(huffMethod)))
```
Comment (guyarb) [inline pkg/network/usm/usm_http2_monitor_test.go:821]:
```go
0b1000000 | len(huffMethod)
```
or
```go
0x80 | len(huffMethod)
```

File: pkg/network/usm/usm_http2_monitor_test.go [hunk 11]
```diff
@@ -800,6 +800,100 @@ func (s *usmHTTP2Suite) TestRawTraffic() {
...
+				}: 1,
+			},
+		},
+		{
+			name: "validate http methods POST and GET methods as literal",
+			// The purpose of this test is to validate that when the methods POST and GET are sent as literal fields
+			// and not indexed, we are still able to capture and process them correctly.
+			messageBuilder: func() [][]byte {
+				getMethodRaw := append([]byte{0x43, 0x03}, []byte("GET")...)
+				postMethodRaw := append([]byte{0x43, 0x04}, []byte("POST")...)
```
Comment (guyarb) [inline pkg/network/usm/usm_http2_monitor_test.go:868]: isn't that covered by the test above?

**Review by guyarb — CHANGES_REQUESTED** [review]

File: pkg/network/usm/usm_http2_monitor_test.go (hunk 10 above)
Comment (guyarb) [reply pkg/network/usm/usm_http2_monitor_test.go:821]: instead of `128+....`

File: pkg/network/usm/usm_http2_monitor_test.go (hunk 11 above)
Comment (amitslavin) [reply pkg/network/usm/usm_http2_monitor_test.go:868]: Actually no, this that validate the plain version and the above one validates the Huffman encoded version.

**Review by guyarb — APPROVED** [review]

**Pull request merged**